File: config/src/main/java/no/nav/melosys/featuretoggle/ByUserIdStrategy.java
package no.nav.melosys.featuretoggle;

import java.util.Arrays;
import java.util.Map;
import java.util.Optional;

import no.finn.unleash.strategy.Strategy;
import no.nav.melosys.sikkerhet.context.SubjectHandler;
import no.nav.melosys.sikkerhet.context.ThreadLocalAccessInfo;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

class ByUserIdStrategy implements Strategy {

    private static final Logger log = LoggerFactory.getLogger(ByUserIdStrategy.class);

    @Override
    public String getName() {
        return "byUserId";
    }

    @Override
    public boolean isEnabled(Map<String, String> parameters) {
        Optional<String> userIDs = Optional.ofNullable(parameters.get("user"));
        if (StringUtils.isEmpty(getLoggedInUserID()) && userIDs.isPresent() && !userIDs.get().isBlank()) {
            log.warn("Finner ikke innlogget bruker i context, mens brukere er registert i unleash. Er sannsynligvis en SED prosess");
        }

        return StringUtils.isNotEmpty(getLoggedInUserID()) && userIDs
            .map(users -> Arrays.asList(users.split(",")))
            .stream().anyMatch(users -> users.contains(getLoggedInUserID()));
    }

    private static String getLoggedInUserID() {
        String userID = null;
        SubjectHandler instance = SubjectHandler.getInstance();
        if (instance != null) {
            userID = instance.getUserID();
        }
        if (userID == null) {
            userID = ThreadLocalAccessInfo.getSaksbehandler();
        }
        return userID;
    }
}

File: config/src/main/kotlin/no/nav/melosys/featuretoggle/ByUserIdStrategy.kt
package no.nav.melosys.featuretoggle

import io.getunleash.strategy.Strategy
import mu.KotlinLogging
import no.nav.melosys.sikkerhet.context.SubjectHandler
import no.nav.melosys.sikkerhet.context.ThreadLocalAccessInfo
import java.util.*

internal class ByUserIdStrategy : Strategy {
    private val log = KotlinLogging.logger { }
    private val uniqueLogMessages = Collections.synchronizedSet(HashSet<String>())

    override fun getName(): String = "byUserId"

    override fun isEnabled(parameters: Map<String, String>): Boolean {
        val userIDs = parameters["user"]
        if (userIDs.isNullOrBlank()) return false

        val loggedInUserID = getLoggedInUserID()
        if (loggedInUserID.isNullOrBlank()) {
            logOnlyFirstMessage(
                "Unleash byUserId Strategy brukes uten innlogget saksbehandler, blir kalt fra:\n${calledFrom()}"
            )
            return false
        }

        return userIDs.split(",").contains(loggedInUserID)
    }

    private fun logOnlyFirstMessage(msg: String) {
        if (uniqueLogMessages.add(msg)) {
            log.warn(msg)
        }
    }

    private fun calledFrom(): String =
        Thread.currentThread().stackTrace.let { stackTraceElements ->
            val element = stackTraceElements.find {
                it.toString().contains(STACK_TRACE_LINE_BEFORE_UNLEASH_IS_ENABLED)
            } ?: return@let "Fant ikke unleash bruk i stacktrace\n" + stackTraceElements.joinToString("\n")
            val indexToLineAfterUnleashIsEnabledCall = stackTraceElements.indexOf(element)
            stackTraceElements[indexToLineAfterUnleashIsEnabledCall + 1].toString()
        }

    private fun getLoggedInUserID(): String? =
        SubjectHandler.getInstance()?.userID ?: ThreadLocalAccessInfo.getSaksbehandler()

    companion object {
        const val STACK_TRACE_LINE_BEFORE_UNLEASH_IS_ENABLED =
            "io.getunleash.DefaultUnleash.isEnabled(DefaultUnleash.java:93"
    }
}
